File: backend/src/main/java/com/pbl5/pbl5/service/StatisticsService.java
package com.pbl5.pbl5.service;

import com.pbl5.pbl5.response.StatisticsResponse;

import java.time.LocalDate;

public interface StatisticsService {
    /**
     * Get dashboard summary statistics including VIP user count, today's page views, and total revenue
     * @return Dashboard summary data
     */
    StatisticsResponse.DashboardSummary getDashboardSummary();

    /**
     * Get revenue chart data for a specified date range
     * @param startDate Optional start date for filtering
     * @param endDate Optional end date for filtering
     * @return Revenue chart data including monthly revenue and subscription counts
     */
    StatisticsResponse.RevenueChart getRevenueData(LocalDate startDate, LocalDate endDate);

    /**
     * Get top manga by view count
     * @param startDate Optional start date for filtering
     * @param endDate Optional end date for filtering
     * @param limit Maximum number of manga to return (default 5)
     * @return List of top manga with view counts
     */
    StatisticsResponse.TopMangaList getTopManga(LocalDate startDate, LocalDate endDate, int limit);
    
    /**
     * Get user statistics with optional date range
     * @param startDate Optional start date for filtering
     * @param endDate Optional end date for filtering
     * @return User statistics data
     */
    StatisticsResponse.UserStats getUserStats(LocalDate startDate, LocalDate endDate);
    
    /**
     * Get payment statistics with optional date range
     * @param startDate Optional start date for filtering
     * @param endDate Optional end date for filtering
     * @return Payment statistics data
     */
    StatisticsResponse.PaymentStats getPaymentStats(LocalDate startDate, LocalDate endDate);
    
    /**
     * Get subscription statistics with optional date range
     * @param startDate Optional start date for filtering
     * @param endDate Optional end date for filtering
     * @return Subscription statistics data
     */
    StatisticsResponse.SubscriptionStats getSubscriptionStats(LocalDate startDate, LocalDate endDate);
}

File: frontend/src/services/statisticsService.js
import api from "./api";

const statisticsService = {
  // Get dashboard summary statistics
  getDashboardSummary: async () => {
    try {
      const response = await api.get("/admin/statistics/dashboard");
      return response.data;
    } catch (error) {
      console.error("Error fetching dashboard summary:", error);
      throw error;
    }
  },

  // Get revenue data with optional date range
  getRevenueData: async (startDate, endDate) => {
    try {
      let url = "/admin/statistics/revenue";
      const params = {};

      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await api.get(url, { params });
      return response.data;
    } catch (error) {
      console.error("Error fetching revenue data:", error);
      throw error;
    }
  },

  // Get top manga with optional date range and limit
  getTopManga: async (startDate, endDate, limit = 5) => {
    try {
      let url = "/admin/statistics/top-manga";
      const params = { limit };

      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await api.get(url, { params });
      return response.data;
    } catch (error) {
      console.error("Error fetching top manga:", error);
      throw error;
    }
  },

  // Get user statistics with optional date range
  getUserStats: async (startDate, endDate) => {
    try {
      let url = "/admin/statistics/users";
      const params = {};

      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await api.get(url, { params });
      return response.data;
    } catch (error) {
      console.error("Error fetching user statistics:", error);
      throw error;
    }
  },

  // Get payment statistics with optional date range
  getPaymentStats: async (startDate, endDate) => {
    try {
      let url = "/admin/statistics/payments";
      const params = {};

      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await api.get(url, { params });
      return response.data;
    } catch (error) {
      console.error("Error fetching payment statistics:", error);
      throw error;
    }
  },

  // Get subscription statistics with optional date range
  getSubscriptionStats: async (startDate, endDate) => {
    try {
      let url = "/admin/statistics/subscriptions";
      const params = {};

      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await api.get(url, { params });
      return response.data;
    } catch (error) {
      console.error("Error fetching subscription statistics:", error);
      throw error;
    }
  },
};

export default statisticsService;
